### ISSUE TITLE
Error w/gradient legends when interactively filtering dataset

### ISSUE THREAD

**Issue by arvind** [vega/vega#1401]
In the following spec (simplified from vega/vega-lite#3865), the gradient legend is driven by a scale with a backing dataset that is interactively filtered via brushing. When interaction begins, the code paths that calculate stops for the gradient legend produce `NaN`s. As a result, the Canvas renderer crashes while the browser throws errors when using the SVG renderer. 

```json
{
  "$schema": "https://vega.github.io/schema/vega/v4.json",
  "autosize": "pad",
  "padding": 5,
  "width": 200,
  "height": 200,
  "style": "cell",
  "data": [
    {"name": "box_store"},
    {"name": "source_0", "url": "data/cars.json"},
    {
      "name": "data_0",
      "source": "source_0",
      "transform": [
        {
          "type": "filter",
          "expr": "!(length(data(\"box_store\"))) || (vlInterval(\"box_store\", datum))"
        }
      ]
    }
  ],
  "signals": [
    {
      "name": "unit",
      "value": {},
      "on": [
        {"events": "mousemove", "update": "isTuple(group()) ? group() : unit"}
      ]
    },
    {
      "name": "box_x",
      "value": [],
      "on": [
        {
          "events": {"source": "scope", "type": "mousedown"},
          "update": "[x(unit), x(unit)]"
        },
        {
          "events": {
            "source": "window",
            "type": "mousemove",
            "consume": true,
            "between": [
              {"source": "scope", "type": "mousedown"},
              {"source": "window", "type": "mouseup"}
            ]
          },
          "update": "[box_x[0], clamp(x(unit), 0, width)]"
        }
      ]
    },
    {
      "name": "box_Horsepower",
      "on": [
        {
          "events": {"signal": "box_x"},
          "update": "box_x[0] === box_x[1] ? null : invert(\"x\", box_x)"
        }
      ]
    },
    {
      "name": "box_tuple",
      "on": [
        {
          "events": [{"signal": "box_Horsepower"}],
          "update": "box_Horsepower ? {unit: \"layer_0\", intervals: [{encoding: \"x\", field: \"Horsepower\", extent: box_Horsepower}]} : null"
        }
      ]
    },
    {
      "name": "box_modify",
      "on": [
        {
          "events": {"signal": "box_tuple"},
          "update": "modify(\"box_store\", box_tuple, true)"
        }
      ]
    }
  ],
  "marks": [
    {
      "name": "box_brush_bg",
      "type": "rect",
      "clip": true,
      "encode": {
        "enter": {"fill": {"value": "#333"}, "fillOpacity": {"value": 0.125}},
        "update": {
          "x": [
            {
              "test": "data(\"box_store\").length && data(\"box_store\")[0].unit === \"layer_0\"",
              "signal": "box_x[0]"
            },
            {"value": 0}
          ],
          "y": [
            {
              "test": "data(\"box_store\").length && data(\"box_store\")[0].unit === \"layer_0\"",
              "value": 0
            },
            {"value": 0}
          ],
          "x2": [
            {
              "test": "data(\"box_store\").length && data(\"box_store\")[0].unit === \"layer_0\"",
              "signal": "box_x[1]"
            },
            {"value": 0}
          ],
          "y2": [
            {
              "test": "data(\"box_store\").length && data(\"box_store\")[0].unit === \"layer_0\"",
              "field": {"group": "height"}
            },
            {"value": 0}
          ]
        }
      }
    },
    {
      "name": "layer_0_marks",
      "type": "symbol",
      "from": {"data": "source_0"},
      "encode": {
        "update": {
          "fill": {"value": "transparent"},
          "stroke": {"value": "lightgray"},
          "x": {"scale": "x", "field": "Horsepower"},
          "y": {"scale": "y", "field": "Acceleration"}
        }
      }
    },
    {
      "name": "layer_1_marks",
      "type": "symbol",
      "from": {"data": "data_0"},
      "encode": {
        "update": {
          "fill": {"value": "transparent"},
          "stroke": {"scale": "color", "field": "Cylinders"},
          "x": {"scale": "x", "field": "Horsepower"},
          "y": {"scale": "y", "field": "Acceleration"}
        }
      }
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "linear",
      "domain": {"data": "source_0", "field": "Horsepower"},
      "range": [0, {"signal": "width"}]
    },
    {
      "name": "y",
      "type": "linear",
      "domain": {"data": "source_0", "field": "Acceleration"},
      "range": [{"signal": "height"}, 0]
    },
    {
      "name": "color",
      "type": "sequential",
      "domain": {"data": "data_0", "field": "Cylinders"},
      "range": "ramp"
    }
  ],
  "legends": [{"stroke": "color", "title": "Cylinders"}]
}
```

**Comment by jheer** [comment]
Thanks. I'm able to reproduce in the online editor, but I don't have any errors when testing locally. I'll need to investigate further...

**Comment by jheer** [comment]
Here's the breakdown:
- Create an interval containing zero points, this causes all records in `data_0` to be filtered out.
- The resulting extent calculation for zero points is `[Infinity, -Infinity]`. This becomes the domain of the `color` scale.
- The gradient generation tries to interpolate between `Infinity` and `-Infinity`, which results in the observed error.

There are a few points that might benefit from intervention:
1. When computing the numeric extent of a field over an empty dataset, is `[Infinity, -Infinity]` a reasonable result, or should we output something else?
2. Should the gradient generation have safeguards against invalid values, or should we only remedy the upstream inputs? (e.g., by addressing point 1...)
3. In this case, what should the desired result be for a gradient over a null domain? Should there be a single answer for all null domains, or should the context (e.g., filtered data) be taken into account? The answer to this question determines not only what Vega should do by default, but what if anything Vega-Lite should be doing as well...

Thoughts?

**Comment by jheer** [comment]
A bit more texture... if the extent calculation instead returns `[undefined, undefined]` for an empty data set, the resulting gradient is:

![image](https://user-images.githubusercontent.com/202799/44432347-8ad84580-a556-11e8-9777-3167dfd0ba37.png)

If the extent calculation returns `[null, null]`, the gradient is similar, but with a singular value of `0` (null coerces to zero) instead of `NaN` (undefined coerces to NaN). In interactive contexts it might be strange for the scale domain to jump out of range of the original data... but I don't have a better default.

**Comment by arvind** [comment]
In prior Vega versions, we used to cache previous scale instances to help with animation. I wonder if a similar strategy could help here? E.g., caching the scale domain/range at initialization time and defaulting to that for gradient generation in the event of invalid extents? In some sense, that would mimic our default configuration of empty selections (as containing all tuples in the dataset).

**Comment by jheer** [comment]
Fixed in vega version 4.3.0.
